lib/internal/worker.js

```diff
@@ -0,0 +1,275 @@
+'use strict';
+
+const Buffer = require('buffer').Buffer;
+
+const {
+  Worker: WorkerImpl,
+  getFromParentChannel,
+  sendToParent,
+  setSerializeWorkerErrorFunction,
+  OOB_NONE,
+  OOB_CUSTOM_OFFSET,
+  OOB_COULD_NOT_SERIALIZE_ERR,
+  OOB_OUT_OF_MEMORY,
+  OOB_ERROR_MESSAGE
+} = process.binding('worker');
+
+setSerializeWorkerErrorFunction(serializeError);
+
+const assert = require('assert');
+const errors = require('internal/errors');
+const { Serializer, Deserializer } = require('v8');
```

Comment:
We need a way to pass binary data fast between isolates - given we can't pass structured data (right?) I think that we need a way to pass buffers and other node builtins fast if we can.

Reply:
The custom Serializer does support Node.js Buffers. Are you saying the performance of going through the serializer might be suboptimal?

Reply:
I'm saying that passing binary data efficiently between workers needs to be fast, it would be amazing to be able to pass objects faster than structured-cloning them but I don't see that happening any time soon.

Reply:
One of the most commonly misunderstood problems people want to solve with workers is "offload parsing a lot of JSON" where the overhead is creating all the objects and not reading the actual JSON and passing it to a worker won't help (since now 2 copies are made per worker because it's deserializaed from JSON to object, then between workers, then to the user).

I don't think we can fix that though.

Reply:
@benjamingr Maybe to clarify: Are you saying the serialization (that corresponds to structured cloning, and afaik is what Chrome implements structured cloning on top of) needs to be fast, or that passing in a Buffer on the one side and getting out another Buffer on the other side needs to be fast?

Anyway:
- I think we might be able to get rid of one (?) of the copy operations along the way when a Buffer is passed in by adding a fast case for “this is just an ArrayBufferView”
- I would also like to add `SharedArrayBuffer`/transferable support, now or in the future

> I don't think we can fix that though.

Yeah.

Reply:
> that passing in a Buffer on the one side and getting out another Buffer on the other side needs to be fast?

That, I think that passing a buffer should be fast (and optimally a binary stream). I have no hopes for anything that requires actually copying JS objects from one heap to another.

Reply:
We won’t be able to get around one of the copies, because any `Buffer` that’s being passed in could be modified in the sending thread afterwards, but I think keeping it at a single copy should work.

Reply:
@benjamingr Yeah. A FAQ/explainer for questions like these should be good, though that might sound too much like a Node.js EP... *sigh*

Reply:
@addaleax I'm fine with modifying a sent buffer `throw`ing or providing a different type.

I'm not too concerned with the API but I definitely think workers should have the capability to pass memory without copies _somehow_.

For example, if I'm using workers to transcode video and I'm doing it at 20Mbps in 4 workers - that means my process is copying 4.8 gigabits of memory every minute which is a lot of overhead.

Reply:
> I'm fine with modifying a sent buffer `throw`ing or providing a different type.

Oh, right. I think we can `Neuter()` array buffers like that (or even have to do that when using them as transferrables?).

> I'm not too concerned with the API but I definitely think workers should have the capability to pass memory without copies _somehow_.

I think that would be doable with SharedArrayBuffers in any case

(to be clear, leaving out support for SABs/transferrables is not a design decision of this PR, it’s a `TODO` of it :D)